File: push/appambit_sdk_push_notifications/android/src/main/kotlin/com/example/appambit_sdk_push_notifications/AppambitSdkPushNotificationsPlugin.kt
package com.example.appambit_sdk_push_notifications

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.os.Handler
import android.os.Looper
import androidx.core.app.ActivityCompat
import com.appambit.sdk.PushKernel
import com.appambit.sdk.PushNotifications
import com.appambit.sdk.models.AppAmbitNotification
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry

class AppambitSdkPushNotificationsPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware,
    PluginRegistry.RequestPermissionsResultListener,
    PluginRegistry.NewIntentListener {

    companion object {
        // Mirror the package-private constants from MessagingService so warm-start
        // intent parsing works without depending on SDK internals.
        private const val ACTION_NOTIFICATION_OPENED = "com.appambit.sdk.NOTIFICATION_OPENED"
        private const val EXTRA_TITLE      = "appambit_title"
        private const val EXTRA_BODY       = "appambit_body"
        private const val EXTRA_COLOR      = "appambit_color"
        private const val EXTRA_ICON       = "appambit_icon"
        private const val EXTRA_IMAGE_URL  = "appambit_image_url"
        private const val EXTRA_DATA_KEYS  = "appambit_data_keys"
        private const val EXTRA_DATA_VALUES = "appambit_data_keys_values"

        private const val PREFS_NAME         = "appambit_push_prefs"
        private const val KEY_HAS_PENDING    = "appambit_push_has_pending"
        private const val KEY_PENDING_ENABLED = "appambit_push_pending_enabled"
    }

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var activity: Activity? = null
    private var pendingResult: Result? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private var listenersInstalled = false

    // Registered as the PushKernel opened-listener in onAttachedToEngine so cold-start
    // taps (handleNotificationOpened fires from onAttachedToActivity, before the Dart
    // entrypoint has run setOpenedListener) are captured. The proxy buffers them until
    // Dart calls "setOpenedListener" and the channel callback is wired (see onMethodCall),
    // at which point the buffered tap is replayed. Wiring the callback here instead would
    // race the Dart side and drop cold-start taps, since invokeMethod has no handler yet.
    private val openedProxy = OpenedProxy()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "appambit_sdk_push_notifications")
        channel.setMethodCallHandler(this)
        PushKernel.setOpenedNotificationListener(openedProxy)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "start" -> {
                // Boot the SDK so it registers its FCM token listener. We use PushKernel.start
                // (not PushNotifications.start) to avoid calling updateConsumer with a hardcoded
                // pushEnabled=true on every launch. The token listener below handles the cases
                // where a new token is needed (e.g. after deleteToken during offline disable).
                PushKernel.start(context)

                // Install a token listener so that when FCM delivers a new token (e.g. after
                // deleteToken was called during an offline disable), the correct enabled state
                // is synced to the backend via the public PushNotifications facade, which reads
                // the stored token internally. pushEnabled=null tells the facade to read the
                // stored enabled state from SharedPrefs.
                PushKernel.setTokenListener(object : PushKernel.TokenListener {
                    override fun onNewToken(token: String) {
                        // PushNotifications.setNotificationsEnabled reads the token from the
                        // SDK's own store — no getCurrentToken() needed here. handleNewToken
                        // deduplicates (if token == currentToken: return), so calling
                        // setNotificationsEnabled here does NOT cause an infinite loop.
                        PushNotifications.setNotificationsEnabled(
                            context,
                            PushKernel.isNotificationsEnabled(context)
                        )
                        clearPendingSync()
                    }
                })

                // Flush any state that couldn't be sent while offline.
                flushPendingSyncIfNeeded()

                installNotificationListenersIfNeeded()
                result.success(null)
            }
            "setNotificationsEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") ?: true
                savePendingSync(enabled)
                if (isNetworkAvailable()) {
                    // Online: use the public facade — it updates SharedPrefs, DB, and backend.
                    PushNotifications.setNotificationsEnabled(context, enabled)
                    clearPendingSync()
                } else {
                    // Offline: only update in-memory / SharedPrefs state so this session is
                    // consistent. Do NOT call PushNotifications (which would write to DB and
                    // contaminate deduplication), leaving the DB at its previous value so that
                    // flushPendingSyncIfNeeded can detect the mismatch on the next online launch.
                    PushKernel.setNotificationsEnabled(context, enabled)
                }
                result.success(null)
            }
            "isNotificationsEnabled" -> {
                result.success(PushKernel.isNotificationsEnabled(context))
            }
            "requestNotificationPermission" -> {
                handleNotificationPermissionRequest(result)
            }
            "hasNotificationPermission" -> {
                val hasPermission = if (Build.VERSION.SDK_INT >= 33) {
                    context.checkSelfPermission(android.Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED
                } else {
                    true
                }
                result.success(hasPermission)
            }
            "setForegroundListener" -> {
                installNotificationListenersIfNeeded()
                result.success(null)
            }
            "setOpenedListener" -> {
                // Dart is now ready to receive opened taps: wire the channel callback,
                // which also flushes any cold-start tap buffered by openedProxy.
                openedProxy.setCallback { notification ->
                    mainHandler.post {
                        channel.invokeMethod("onOpenedNotification", PushPayloadMapper.payloadOf(notification))
                    }
                }
                installNotificationListenersIfNeeded()
                result.success(null)
            }
            "setBackgroundHandler" -> {
                val args = call.arguments as? Map<*, *>
                val dispatcherHandle = (args?.get("dispatcherHandle") as? Number)?.toLong()
                val handlerHandle = (args?.get("handlerHandle") as? Number)?.toLong()
                if (dispatcherHandle == null || handlerHandle == null) {
                    result.error(
                        "INVALID_ARGS",
                        "setBackgroundHandler requires dispatcherHandle and handlerHandle.",
                        null
                    )
                } else {
                    AppambitFlutterPushExtension.saveHandles(context, dispatcherHandle, handlerHandle)
                    result.success(null)
                }
            }
            else -> result.notImplemented()
        }
    }

    // MARK: - Pending sync

    private fun savePendingSync(enabled: Boolean) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).edit()
            .putBoolean(KEY_HAS_PENDING, true)
            .putBoolean(KEY_PENDING_ENABLED, enabled)
            .apply()
    }

    private fun clearPendingSync() {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).edit()
            .remove(KEY_HAS_PENDING)
            .remove(KEY_PENDING_ENABLED)
            .apply()
    }

    // Flushes any consumer update that was deferred because the device was offline.
    //
    // For disable (enabled=false): PushNotifications.setNotificationsEnabled reads the
    // stored token from DB (the last valid FCM token before deleteToken was called).
    //
    // For enable (enabled=true) with no token yet: skipped here; the token listener
    // installed in "start" handles it when FCM delivers the new token, which happens
    // on the next online launch after deleteToken was called during offline disable.
    private fun flushPendingSyncIfNeeded() {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        if (!prefs.getBoolean(KEY_HAS_PENDING, false)) return
        val enabled = prefs.getBoolean(KEY_PENDING_ENABLED, true)

        if (!isNetworkAvailable()) return

        if (!enabled) {
            // Disable: the stored FCM token in the DB is still valid (deleteToken only
            // clears the in-memory token, not the DB copy). The public facade reads it.
            PushNotifications.setNotificationsEnabled(context, false)
            clearPendingSync()
        }
        // Enable with no token: token listener fires when FCM delivers the new token.
        // PushKernel.isNotificationsEnabled will return true (set by PushKernel.setNotificationsEnabled
        // during the offline enable call), so onNewToken will call setNotificationsEnabled(true).
    }

    // MARK: - Network

    private fun isNetworkAvailable(): Boolean {
        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            ?: return false
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            cm.getNetworkCapabilities(cm.activeNetwork)
                ?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) == true
        } else {
            @Suppress("DEPRECATION")
            cm.activeNetworkInfo?.isConnected == true
        }
    }

    // MARK: - Notification listeners

    private fun installNotificationListenersIfNeeded() {
        if (listenersInstalled) return
        listenersInstalled = true

        PushKernel.setForegroundNotificationListener(object : PushKernel.ForegroundNotificationListener {
            override fun onForegroundNotificationReceived(notification: AppAmbitNotification) {
                mainHandler.post {
                    channel.invokeMethod("onForegroundNotification", PushPayloadMapper.payloadOf(notification))
                }
            }
        })
        // openedProxy is already installed on PushKernel in onAttachedToEngine.
    }

    // MARK: - Permission

    private fun handleNotificationPermissionRequest(result: Result) {
        val currentActivity = activity
        if (currentActivity == null) {
            result.error("NO_ACTIVITY", "The current Activity is not available", null)
            return
        }

        if (Build.VERSION.SDK_INT < 33) {
            result.success(true)
            return
        }

        if (currentActivity.checkSelfPermission(android.Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED) {
            result.success(true)
        } else {
            pendingResult = result
            ActivityCompat.requestPermissions(
                currentActivity,
                arrayOf(android.Manifest.permission.POST_NOTIFICATIONS),
                101
            )
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ): Boolean {
        if (requestCode == 101) {
            val granted = grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED
            pendingResult?.success(granted)
            pendingResult = null
            return true
        }
        return false
    }

    // MARK: - Flutter lifecycle

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addRequestPermissionsResultListener(this)
        binding.addOnNewIntentListener(this)
        // Cold-start tap: the activity was launched from a notification.
        // openedProxy buffers the result if PushKernel.start() hasn't run yet.
        binding.activity.intent?.let {
            PushKernel.handleNotificationOpened(binding.activity, it)
        }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addRequestPermissionsResultListener(this)
        binding.addOnNewIntentListener(this)
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onNewIntent(intent: Intent): Boolean {
        // Warm-start tap: the activity is already running and a notification
        // brought it to the foreground. Bypass the PushKernel → openedProxy chain
        // and invoke the channel directly so the delivery is not affected by any
        // intermediate listener state during the background→foreground transition.
        if (intent.action == ACTION_NOTIFICATION_OPENED) {
            val notification = notificationFromIntent(intent)
            mainHandler.post {
                channel.invokeMethod("onOpenedNotification", PushPayloadMapper.payloadOf(notification))
            }
        }
        return false
    }

    private fun notificationFromIntent(intent: Intent): AppAmbitNotification {
        val keys   = intent.getStringArrayExtra(EXTRA_DATA_KEYS)
        val values = intent.getStringArrayExtra(EXTRA_DATA_VALUES)
        val data   = mutableMapOf<String, String>()
        if (keys != null && values != null) {
            for (i in keys.indices) if (i < values.size) data[keys[i]] = values[i]
        }
        return AppAmbitNotification(
            intent.getStringExtra(EXTRA_TITLE),
            intent.getStringExtra(EXTRA_BODY),
            intent.getStringExtra(EXTRA_COLOR),
            intent.getStringExtra(EXTRA_ICON),
            intent.getStringExtra(EXTRA_IMAGE_URL),
            data
        )
    }

    // MARK: - Opened notification buffering proxy

    private class OpenedProxy : PushKernel.OpenedNotificationListener {
        @Volatile private var callback: ((AppAmbitNotification) -> Unit)? = null
        @Volatile private var buffered: AppAmbitNotification? = null

        fun setCallback(cb: (AppAmbitNotification) -> Unit) {
            callback = cb
            val buf = buffered
            if (buf != null) {
                buffered = null
                cb(buf)
            }
        }

        override fun onOpenedNotification(notification: AppAmbitNotification) {
            val cb = callback
            if (cb != null) {
                cb(notification)
            } else {
                buffered = notification
            }
        }
    }
}

File: push/appambit_sdk_push_notifications/ios/Classes/AppambitSdkPushNotificationsPlugin.swift
import Flutter
import UIKit
import Network
import UserNotifications
import AppAmbit
import AppAmbitPushNotifications

public class AppambitSdkPushNotificationsPlugin: NSObject, FlutterPlugin {

    private var channel: FlutterMethodChannel?
    private var listenerInstalled = false

    // MARK: - Pending sync keys
    private static let kHasPending = "appambit_push_has_pending"
    private static let kPendingEnabled = "appambit_push_pending_enabled"

    private static func savePendingSync(enabled: Bool) {
        let d = UserDefaults.standard
        d.set(true, forKey: kHasPending)
        d.set(enabled, forKey: kPendingEnabled)
    }

    private static func clearPendingSync() {
        let d = UserDefaults.standard
        d.removeObject(forKey: kHasPending)
        d.removeObject(forKey: kPendingEnabled)
    }

    /// Called from `case "start"` — always after AppAmbit is initialized
    /// (Dart calls AppAmbitSdk.start before PushNotificationsSdk.start).
    ///
    /// Root cause: APNs delivers the token fast on physical devices (~50-100 ms).
    /// TokenListenerImpl.sync fires before AppAmbit is ready, retries once at
    /// +0.5 s, then drops the token. When Dart's start() later replaces the
    /// token listener, handleNewToken's `guard token != currentToken` prevents
    /// re-delivery, so ConsumerService is never called for the pending state.
    ///
    /// Fix: at Dart-start time AppAmbit IS ready. Wait for the APNs token to be
    /// in memory, then call setNotificationsEnabled directly — ConsumerService
    /// gets a valid token and the correct enabled state.
    private static func flushPendingSyncIfNeeded() {
        guard UserDefaults.standard.bool(forKey: kHasPending) else { return }
        let enabled = UserDefaults.standard.bool(forKey: kPendingEnabled)
        // Fire-and-forget — never blocks start().
        waitForTokenThenSync(enabled: enabled, attempt: 0)
    }

    private static func waitForTokenThenSync(enabled: Bool, attempt: Int) {
        // Give up after 5 s (10 × 0.5 s). If no token by then the device is
        // likely offline; the flag stays for the next launch.
        guard attempt < 10 else { return }

        let token = PushKernel.getCurrentToken() ?? ""
        guard !token.isEmpty else {
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                waitForTokenThenSync(enabled: enabled, attempt: attempt + 1)
            }
            return
        }

        // Token is in memory — AppAmbit is already initialized (see caller).
        // Check network before hitting ConsumerService.
        let monitor = NWPathMonitor()
        monitor.pathUpdateHandler = { path in
            monitor.cancel()
            guard path.status == .satisfied else { return }
            DispatchQueue.main.async {
                PushNotifications.setNotificationsEnabled(enabled)
                clearPendingSync()
            }
        }
        monitor.start(queue: DispatchQueue(label: "appambit.netcheck"))
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "appambit_sdk_push_notifications",
            binaryMessenger: registrar.messenger()
        )
        let instance = AppambitSdkPushNotificationsPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)

        // Bootstrap the iOS SDK during `application(_:didFinishLaunchingWithOptions:)`
        // so the UNUserNotificationCenter delegate is installed in time for iOS
        // to deliver a cold-start tap (`didReceive` for `UNNotificationResponse`).
        // We deliberately do NOT install the plugin's notification listener
        // here — that happens later when Dart calls `setOpenedListener`. In the
        // meantime the SDK queues any tap-driven notification internally and
        // replays it once the listener is registered, by which point Dart's
        // `setMethodCallHandler` is also set so the channel message reaches the
        // user's callback. `start()` is idempotent so calling it again from
        // Dart is harmless.
        PushNotifications.start()
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":
            Self.flushPendingSyncIfNeeded()
            PushNotifications.start()
            installNotificationListenerIfNeeded()
            // If the APNs token arrived before AppAmbit was ready (TokenListenerImpl dropped it),
            // sync it now — AppAmbit.start() has completed so consumerId is set.
            let token = PushKernel.getCurrentToken() ?? ""
            if !token.isEmpty {
                ConsumerService.shared.updateConsumer(
                    deviceToken: token,
                    pushEnabled: PushKernel.isNotificationsEnabled()
                )
            }
            result(nil)

        case "setNotificationsEnabled":
            let args = call.arguments as? [String: Any]
            let enabled = args?["enabled"] as? Bool ?? true
            Self.savePendingSync(enabled: enabled)
            // Update in-memory / UserDefaults state immediately so this session is consistent.
            PushKernel.setNotificationsEnabled(enabled)
            // Only write to the DB + hit the backend when online. If offline, the DB stays at
            // its previous value — flushPendingSyncIfNeeded needs that mismatch on the next
            // online launch to bypass ConsumerService deduplication and actually reach the backend.
            let setEnabled = enabled
            let setToken = PushKernel.getCurrentToken()
            let setMonitor = NWPathMonitor()
            setMonitor.pathUpdateHandler = { path in
                setMonitor.cancel()
                guard path.status == .satisfied else { return }
                DispatchQueue.main.async {
                    ConsumerService.shared.updateConsumer(deviceToken: setToken, pushEnabled: setEnabled)
                    Self.clearPendingSync()
                }
            }
            setMonitor.start(queue: DispatchQueue(label: "appambit.setNotif"))
            result(nil)

        case "isNotificationsEnabled":
            result(PushNotifications.isNotificationsEnabled())

        case "requestNotificationPermission":
            PushNotifications.requestNotificationPermission { granted in
                DispatchQueue.main.async {
                    if granted {
                        Self.savePendingSync(enabled: true)
                        // PushKernel already called setNotificationsEnabled(true) internally.
                        // Only write to DB + backend when online (same reason as setNotificationsEnabled).
                        let permToken = PushKernel.getCurrentToken()
                        let permMonitor = NWPathMonitor()
                        permMonitor.pathUpdateHandler = { path in
                            permMonitor.cancel()
                            guard path.status == .satisfied else { return }
                            DispatchQueue.main.async {
                                ConsumerService.shared.updateConsumer(deviceToken: permToken, pushEnabled: true)
                                Self.clearPendingSync()
                            }
                        }
                        permMonitor.start(queue: DispatchQueue(label: "appambit.reqPerm"))
                    }
                    result(granted)
                }
            }

        case "hasNotificationPermission":
            result(PushNotifications.hasNotificationPermission())

        case "setForegroundListener", "setOpenedListener":
            installNotificationListenerIfNeeded()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Notification listener

    private func installNotificationListenerIfNeeded() {
        guard !listenerInstalled else { return }
        listenerInstalled = true

        PushNotifications.setNotificationListener { [weak self] (userInfo: [AnyHashable: Any], state: PushNotificationState) in
            guard let self = self, let channel = self.channel else { return }
            let payload = Self.parsePayload(userInfo)
            let method = (state == .foreground) ? "onForegroundNotification" : "onOpenedNotification"
            DispatchQueue.main.async {
                channel.invokeMethod(method, arguments: payload)
            }
        }
    }

    // MARK: - Payload mapping

    /// Converts an APNs `userInfo` dictionary into the Dart `PushNotificationData`
    /// map shape. Top-level custom keys are forwarded in `data`; the standard
    /// `aps` fields are surfaced under `ios`. The raw `aps` dictionary is not
    /// copied into `data`.
    private static func parsePayload(_ userInfo: [AnyHashable: Any]) -> [String: Any] {
        let aps = userInfo["aps"] as? [String: Any]
        let alert = aps?["alert"] as? [String: Any]

        let title = alert?["title"] as? String
        let body = alert?["body"] as? String
        let imageUrl = userInfo["image"] as? String

        var data: [String: String] = [:]
        for (key, value) in userInfo {
            guard let k = key as? String, k != "aps", k != "image" else { continue }
            data[k] = "\(value)"
        }

        var ios: [String: Any] = [:]
        if let badge = aps?["badge"] as? Int { ios["badge"] = badge }
        if let sound = aps?["sound"] as? String { ios["sound"] = sound }
        if let category = aps?["category"] as? String { ios["category"] = category }
        if let threadId = aps?["thread-id"] as? String { ios["threadId"] = threadId }

        var payload: [String: Any] = [:]
        if let title { payload["title"] = title }
        if let body { payload["body"] = body }
        if let imageUrl { payload["imageUrl"] = imageUrl }
        if !data.isEmpty { payload["data"] = data }
        if !ios.isEmpty { payload["ios"] = ios }
        return payload
    }
}
